### PR TITLE
Fix slow compilation when using std.regex

### PR DESCRIPTION
**Pull request by DmitryOlshansky**
Taking this short program as an example:
```d
import std.regex, std.stdio;

void main(string[] argv) {
	auto re = regex(argv[1]);
	auto matches = matchFirst(argv[2], re);
	writeln(matches);
}
```

Before:
```
time dmd re.d

real	0m2,404s
user	0m2,205s
sys	0m0,200s
```
After:
```
time dmd re.d

real	0m1,198s
user	0m1,103s
sys	0m0,096s
```

I still think the compile time is too big, maybe there is another CTFE-heavy construction hiding in there.

### CODE REVIEW

File: std/parallelism.d [hunk 1]
```diff
@@ -4914,3 +4914,10 @@ version (parallelismStressTest)
     try foreach (rnd; rndGen.parallel) break;
     catch (ParallelForeachError e) {}
 }
+
+nothrow pure @safe unittest
+{
+    // Parallelism (was broken by inferred return type "immutable int")
+    import std.algorithm.comparison;
+    auto t = task!cmp("foo", "bar");
+}
```
Comment (dkorpel) [inline std/parallelism.d:4923]: Moving this test seems unrelated

File: std/regex/internal/ir.d [hunk 2]
```diff
@@ -49,10 +49,11 @@ CharMatcher[CodepointSet] matcherCache;
     }
 }
 
-@property ref wordMatcher()()
+static immutable CharMatcher wordMatcher;
+
+shared static this()
```
Comment (dkorpel) [inline std/regex/internal/ir.d:54]: This would add runtime to every D program, even if it's not using std.regex.

File: std/parallelism.d (hunk 1 above)
Comment (DmitryOlshansky) [reply std/parallelism.d:4923]: This has to do with nice circular dependency problem - if std.regex.ir has module ctor then during unittest compilation a cycle between std.algorithm.comparison and std.parallelism is hit. I'd wish we didn't consider unittests to be part of the module and have them compiled somehow separately but it is what it is. The simplest path out of this hell was to move test to more heavily dependent module - std.parallelism.

File: std/regex/internal/ir.d (hunk 2 above)
Comment (DmitryOlshansky) [reply std/regex/internal/ir.d:54]: From my measurements it's about 1ms and I think you still need to import std.regex to get it to run. If somebody has cheap lazy initialization for global immutables I'd gladly take it.

Comment (rikkimax) [reply std/regex/internal/ir.d:54]: Dmitry is right, that shouldn't be an issue.

File: std/regex/internal/ir.d [hunk 3]
```diff
@@ -971,6 +972,13 @@ struct CharMatcher {
         trie = makeTrie(set);
     }
 
+    immutable this(CodepointSet set)
+    {
+        auto asciiSet = set & unicode.ASCII;
+        ascii = BitTable(asciiSet);
+        trie = makeTrie(set);
```
Comment (RazvanN7) [inline std/regex/internal/ir.d:979]: This trips the tester: `std/regex/internal/ir.d(979): Error: cannot implicitly convert expression `codepointSetTrie(((InversionList!(GcPolicy) __copytmp5388 = (__copytmp5388 = set).__fieldPostblit();) , __copytmp5388))` of type `Trie!(BitPacked!(bool, 1u), dchar, 1114112u, sliceBits!(8u, 21u), sliceBits!(0u, 8u))` to `immutable(Trie!(BitPacked!(bool, 1u), dchar, 1114112u, sliceBits!(8u, 21u), sliceBits!(0u, 8u)))`

Comment (DmitryOlshansky) [reply std/regex/internal/ir.d:979]: It’s all 32-bit platforms. This is strange the only thing that should change is ulong -> uint in the internal structure of a trie. Will see on the weekend if I can figure it out

**Comment by DmitryOlshansky** [comment]
@RazvanN7 I don't get cirrus-ci failing on some UDA test in std.traits that shouldn't be affected by my code at all. Any ideas?

File: std/regex/internal/ir.d (hunk 3 above)
Comment (DmitryOlshansky) [reply std/regex/internal/ir.d:979]: I don't know what I'm doing but I managed to sidestep this problem by passing CodepointSet by const ref.

**Comment by RazvanN7** [comment]
You should probably rebase to the latest version of phobos, that will most likely fix it.

**Comment by DmitryOlshansky** [comment]
In general I approve this pr

**Comment by LightBender** [comment]
PR closed as stalled. This appears to have been addressed by a later PR. If you wish to resurrect it you are welcome to do so.
